src/nvim/lua/converter.c

```diff
@@ -1324,7 +1325,27 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
     if (field->type == kObjectTypeNil) {
       *(Object *)mem = nlua_pop_Object(L, true, err);
     } else if (field->type == kObjectTypeInteger) {
-      *(Integer *)mem = nlua_pop_Integer(L, err);
+      if (field->is_hlgroup) {
+        if (lua_type(L, -1) == LUA_TSTRING) {
+          String name = nlua_pop_String(L, err);
+          int hl_id = 0;
+          if (!ERROR_SET(err) && name.size > 0) {
+            hl_id = syn_check_group(name.data, name.size);
+          }
+          api_free_string(name);
+          *(Integer *)mem = hl_id;
+        } else if (lua_type(L, -1) == LUA_TNUMBER) {
+          *(Integer *)mem = nlua_pop_Integer(L, err);
```

Comment:
we can just do this in an `else`, after we checked for String, `nlua_pop_Integer` will do the rest.